### ISSUE TITLE
Compare Axolotl Qwen vs MBridge Qwen vs MBridge Nemotron Super LoRA SFT performance

### ISSUE THREAD

**Issue by dnola-nvidia** [NVIDIA-NeMo/Megatron-Bridge#4214]
### User problem

As title suggests, MBridge is considerably slower than Axolotl for us. Lack of HSDP Support could be a key contributor, as turning that off in Axolotl makes it go 6x slower - and comparing HSDP off with HSDP off MBridge would be faster. But regardless of how we get there, if we want to adopt MBridge it needs to show a performance advantage over our existing stack - and we can't move forward with FM training without this as training is unacceptably slow.

MBridge with 64 GPUs takes over 50 hours to train. Axolotl with half as many GPUs takes closer to 20.

Config wise we are using:

| Recipe | `megatron.bridge.recipes.nemotronh.nemotron_3_super.nemotron_3_super_peft_config` |
| Dataset | SQuAD (recipe default) |
| Sequence length | 8192 |
| Global batch size | 512 |
| Micro batch size | 4 |
| Gradient accumulation | 16 (= 512 / (mbs=4 × DP=8)) |
| **TP** | **4** |
| **CP** | **2** |
| **PP** | **1** |
| **EP** | **8** |
| sequence_parallel | True |
| LoRA target modules | `linear_qkv,linear_proj,linear_fc1,linear_fc2,in_proj,out_proj` |
| LoRA r / alpha | 64 / 128 |
| LR | recipe default (1e-4 for peft) |
| LR warmup | 100 iters |
| Train iters | 200 (short) |
| RECOMPUTE | 1 (selective activation recompute) |
| Distributed optimizer | False (LoRA path) |
| Resources | 8 nodes × 8 B200 = 64 GPUs |

### Desired outcome

MBridge w/ Nemotron Super to train faster than our existing stack.

### Alternatives or workarounds considered

Here is the Axolotl Config we are comparing with:

[config.yaml](https://github.com/user-attachments/files/28724505/config.yaml)

### Affected area

area:perf

### Urgency / use case

Blocking current work

### Environment

| Model | `nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-BF16` |
| Container | `nvcr.io/nvidia/nemo:26.04` (with bundled Megatron-Bridge r0.4.0) |

**Comment by dnola-nvidia** [comment]
[finetune_squad.py](https://github.com/user-attachments/files/28724835/finetune_squad.py)
[sky_launch.yaml](https://github.com/user-attachments/files/28724836/sky_launch.yaml)

More reproduction details - this should be enough to reproduce.

Let us know if you need other details - thanks!

**Comment by yaoyu-33** [comment]
Measured number "30s for bridge with 64GPUs, 12s for axolotl with 32GPUs"

**Comment by yaoyu-33** [comment]
Thanks for the extra repro details. One config sanity check: the attached Axolotl config appears to be for `Qwen/Qwen3.5-122B-A10B` (for example `_name_or_path`, `base_model_name_or_path`, and `model_type=qwen3_5_moe`), while the Bridge repro table is for `nvidia/NVIDIA-Nemotron-3-Super-120B-A12B-BF16`.

So the current `30s/step Bridge on 64 GPUs` vs `12s/step Axolotl on 32 GPUs` comparison does not look apples-to-apples yet. I also see several other large deltas in the attached configs: Axolotl has `sample_packing=true`, FSDP/HSDP-style sharding, per-device train batch size 2, LoRA r/alpha 256/512 with broader target modules/parameters; the Bridge launcher uses TP=4/CP=2/EP=8, MBS=4/GBS=512, LoRA r/alpha 64/128, and disables distributed optimizer/DDP overlap.

We will still try to reproduce the Bridge timing, but before interpreting it as a Bridge regression we should align the effective configs and compare tokens/sec under the same model/data/packing setup. Could you confirm whether the attached Axolotl config is the intended baseline for this issue, or share the Axolotl config used for Nemotron-3-Super if different?

**Comment by dnola-nvidia** [comment]
We were hoping to move from Axolotl+Qwen to Nemotron Super+MBridge - so the Qwen config was the intended baseline. We can look into Nemotron super via Axolotl but it’s possible that could be a challenge - so ideally we can optimize MBridge independently of collecting that info. 

They have similar param counts as MoEs, so the substantially worse perf is not what we expected - Nemotron and MBridge would hopefully be a performance win

**Comment by dnola-nvidia** [comment]
To add, distributed optimizer and sample packing in MBridge make it slower - the config above is the fastest we have found.

**Comment by dnola-nvidia** [comment]
Update - it’s still on a different batch size so step times aren’t directly comparable, but the numbers for this 64 GPU run are a better target. 

qwen 122b lora training run on 64 B200s:
lora rank=128, batch size per gpu =8, tensor parallel=1sharding across 8 GPUs, DDP across 8 nodesmax seq length=8192.~15B total tokens (5B trainable). 

It took 30 hours vs 50 for Nemotron. 

So a closer gap end to end - step times wouldn’t be directly comparable but we can see the improvement ratio we need is order of ~1.67x 

So that is going to be a better target

**Comment by dnola-nvidia** [comment]
More details on the 64 GPU Axolotl run:
```
# Axolotl: Qwen3.5-122B-A10B LoRA SFT
# Usage:
#   sky launch sky/train_multinode.yaml --env-file .env --num-nodes 4 \
#     --env TRAIN_CONFIG=configs/axolotl/lora_sft_122b_conv_note_trace.yaml

plugins:
  - axolotl.integrations.cut_cross_entropy.CutCrossEntropyPlugin
  - axolotl.integrations.liger.LigerPlugin
liger_rope: true
liger_rms_norm: true
liger_swiglu: true
strict: false

base_model: Qwen/Qwen3.5-122B-A10B

tokenizer_type: AutoTokenizer

load_in_8bit: false
load_in_4bit: false

# ---------- LoRA ----------
adapter: lora
lora_r: 128
lora_alpha: 256
lora_dropout: 0.0           # must be 0 when lora_target_parameters is set
lora_target_linear: true
lora_target_parameters:
  - mlp.experts.gate_up_proj  # 3D nn.Parameter — not reachable via lora_target_linear
  - mlp.experts.down_proj
lora_fan_in_fan_out: false

# ---------- Datasets ----------
datasets:
  - path: 
    type: chat_template
    ds_type: json
    field_messages: messages
    roles_to_train:
      - assistant
    message_property_mappings:
      role: role
      content: content
    chat_template_kwargs:
      enable_thinking: false

  - path: 
    type: chat_template
    ds_type: json
    field_messages: messages
    roles_to_train:
      - assistant
    message_property_mappings:
      role: role
      content: content
    chat_template_kwargs:
      enable_thinking: false

  - path: 
    type: chat_template
    ds_type: json
    field_messages: messages
    roles_to_train:
      - assistant
    message_property_mappings:
      role: role
      content: content
    chat_template_kwargs:
      enable_thinking: false

  - path: 
    type: chat_template
    ds_type: json
    field_messages: messages
    roles_to_train:
      - assistant
    message_property_mappings:
      role: role
      content: content
    chat_template_kwargs:
      enable_thinking: false

  - path: 
    type: chat_template
    ds_type: json
    field_messages: messages
    roles_to_train:
      - assistant
    message_property_mappings:
      role: role
      content: content
    chat_template_kwargs:
      enable_thinking: false


val_set_size: 0.0
dataset_prepared_path:

# ---------- Sequence / packing ----------
sequence_len: 8192
sample_packing: true
pad_to_sequence_len: false
eval_sample_packing: false

# ---------- Training hyperparameters ----------
output_dir: 

# Global batch = 64 GPUs × micro_batch 8 × grad_accum 2 = 1024
micro_batch_size: 8
gradient_accumulation_steps: 2
num_epochs: 1

optimizer: adamw_torch_fused
lr_scheduler: cosine
learning_rate: 2.0e-4
weight_decay: 0.0
adam_beta1: 0.9
adam_beta2: 0.95
max_grad_norm: 1.0
warmup_ratio: 0.03

bf16: auto
tf32: true

flash_attention: true

# ---------- FSDP2 ----------
fsdp_version: 2
fsdp_config:
  reshard_after_forward: true
  auto_wrap_policy: TRANSFORMER_BASED_WRAP
  transformer_layer_cls_to_wrap: Qwen3_5MoeDecoderLayer
  state_dict_type: FULL_STATE_DICT
  cpu_ram_efficient_loading: true
  activation_checkpointing: true

dp_shard_size: 8
dp_replicate_size: 8
tensor_parallel_size: 1

# ---------- Saving ----------
save_strategy: steps
save_steps: 100
save_total_limit: 3

# ---------- Logging ----------
logging_steps: 10
logging_first_step: true

# W&B
wandb_project: qwen3-medical
wandb_name: qwen3.5-122b-conv-note-trace-sft-lora-experts-05_21_2026

# ---------- Misc ----------
seed: 42
special_tokens:
  pad_token: <|endoftext|>
```

**Comment by cuichenx** [comment]
Hi @dnola-nvidia could you try applying this Megatron Core patch? https://github.com/NVIDIA/Megatron-LM/pull/5092
If you see this bug only with packed sequence, and only with PEFT (not full parameter SFT), then 5092 could be the solution. 
Happy to provide more details if you'd like.

**Comment by dnola-nvidia** [comment]
Thank you!

Packed sequence is currently off because it is faster in megatron. Are you saying that turning it on and using this patch would likely be better?

**Comment by cuichenx** [comment]
Yes, 5092 fixes an issue which is only observed when packed sequence is on. 
After 5092, you should see that packed sequence run speed is more comparable to unpacked. (however packing passes more effective tokens to reduce the number of gradient accumulation steps so overall it should converge faster than the unpacked case)

**Comment by dnola-nvidia** [comment]
After applying that patch and switching to packing, the time per epoch ends up even slower unfortunately, even accounting for higher token throughput from packing - so the PR doesn’t help. 

The issue there seems to be that packing ends up needing MBS=1 to do, turning packing off allows for MBS 4

Other ideas for how we can close the gap in performance?

**Comment by yaoyu-33** [comment]
Thanks for trying the patch. One clarification on packed-sequence semantics in Bridge: for offline packed SFT, `micro_batch_size=1` is expected. In that mode, MBS counts **packed sequences**, not original training examples. One packed sequence can contain multiple original examples up to `packed_sequence_size`, while preserving attention/loss boundaries through packed metadata. So `packed MBS=1` vs `unpacked MBS=4` is not a direct comparison of the effective batch size or amount of useful work per step.

The docs that describe this are:

- Packed sequence overview: https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/docs/training/packed-sequences.md
- Performance guide, sequence packing section: https://github.com/NVIDIA-NeMo/Megatron-Bridge/blob/main/docs/performance-guide.md#sequence-packing-for-performant-fine-tuning

The more useful comparison is effective non-padding tokens/sec, plus the effective tokens/examples per optimizer step after packing. For a packed run, please also check/report packing utilization, `packed_sequence_size`, gradient accumulation, global batch size, and the number of original examples per pack. If packing reduces padding but total epoch time is still worse, then either the MCore patch is not addressing the bottleneck in this setup, or there is still a config mismatch.

@cuichenx could you sanity-check whether NVIDIA/Megatron-LM#5092 is expected to help this exact case? My current read is that it is only expected to help if the primary slowdown is the packed PEFT path itself. If the packed run is still slower after accounting for useful-token throughput, we probably need to isolate the config/model differences first.

Suggested next step: before comparing Axolotl Qwen directly against Bridge Nemotron Super, let's do a Qwen-vs-Qwen side-by-side first with the same model family, dataset, sequence length, LoRA target/rank as closely as possible, and effective tokens per optimizer step. After that, we can compare Bridge Qwen vs Bridge Nemotron Super to isolate recipe/model differences separately.

**Comment by cuichenx** [comment]
No need to compare tokens/sec. All you need to do is compare packing + mbs1 with and without 5092. If it's slower with 5092, then perhaps its not the same issue I encountered before. A good next step is to enable profiling to find out where exactly is the slowdown. You can send us the logs to analyze if needed.
```
  profiling.use_nsys_profiler=true \
  profiling.profile_step_start=10 \
  profiling.profile_step_end=20 \
  profiling.profile_ranks=[0]
```

**Comment by suiyoubi** [comment]
Hi @dnola-nvidia  were you able to do some profiling and identify exactly where is the slowdown ? happy to take a look at the profiling logs if you have them.
